src/dns.rs

```diff
@@ -28,22 +30,21 @@ pub enum AddrType {
 pub trait Dns {
 	/// The type returned when we have an error
 	type Error: core::fmt::Debug;
-	/// An iterator of IP addresses.
-	type IpAddrIter: Iterator<Item = IpAddr>;
 
 	/// Resolves the IP addresses associated with a host, given its hostname and a desired
 	/// address record type to look for.
-	fn get_host_by_name(
+	fn get_hosts_by_name<'a>(
 		&mut self,
 		hostname: &str,
 		addr_type: AddrType,
-	) -> nb::Result<Self::IpAddrIter, Self::Error>;
+		outputs: &'a mut [MaybeUninit<IpAddr>],
+	) -> Result<&'a [IpAddr], Self::Error>;
```

Comment:
I think we can just tie the lifetime of `'a` to that of `self`. That way, any struct that `impl Dns` can just borrow out internal storage. E.g.

```rust
struct MyDns {
   storage: [IpAddr; 5],
}

impl Dns for MyDns {
    fn get_hosts_by_name<'s>(&'s mut self, hostname: &str, addr_type: AddrType) -> Results<&'s [IpAddr], Self::Error> {
        &self.storage[..]
    }
}
```

Or something similar? I don't think the user should need to pass in a storage buffer - that's a very c-like paradigm that I don't think is required with the ownership semantics Rust provides

Reply:
Can we add a const-generic to this function argument to allow a list of addresses to be returned?

```rust
fn get_hosts_by_name<const T: usize>(&mut self, hostname: &str, addr_type: AddrType) -> Result<heapless::Vec<IpAddr, T>, Self::Error>;
```

Reply:
Yeah, that'd be fine. I think that'd optimize to pretty much the same thing as passing in a slice.